src/action.rs

```diff
@@ -0,0 +1,140 @@
+//! Action builder types.
+
+mod drop;
+mod list_databases;
+mod perf;
+mod session;
+mod shutdown;
+mod watch;
+
+pub use drop::{DropDatabase, DropDatabaseFuture};
+pub use list_databases::{ListDatabases, ListSpecificationsFuture};
+pub use perf::{WarmConnectionPool, WarmConnectionPoolFuture};
+pub use session::{StartSession, StartSessionFuture};
+pub use shutdown::{Shutdown, ShutdownFuture};
+pub use watch::{Watch, WatchExplicitSessionFuture, WatchImplicitSessionFuture};
+
+macro_rules! option_setters {
+    (
+        $opt_field:ident: $opt_field_ty:ty;
+        $(
+            $(#[$($attrss:tt)*])*
+            $opt_name:ident: $opt_ty:ty,
+        )+
+    ) => {
+        fn options(&mut self) -> &mut $opt_field_ty {
+            self.$opt_field.get_or_insert_with(<$opt_field_ty>::default)
+        }
+
+        #[cfg(test)]
+        #[allow(dead_code)]
+        pub(crate) fn with_options(mut self, value: impl Into<Option<$opt_field_ty>>) -> Self {
+            self.options = value.into();
+            self
+        }
+
+        $(
+            $(#[$($attrss)*])*
+            pub fn $opt_name(mut self, value: $opt_ty) -> Self {
+                self.options().$opt_name = Some(value);
+                self
+            }
+        )+
+    };
+}
+use option_setters;
+
+/// Generates:
+/// * an `IntoFuture` executing the given method body
+/// * an opaque wrapper type for the future in case we want to do something more fancy than
+///   BoxFuture.
+/// * a `run` method for sync execution, optionally with a wrapper function
+macro_rules! action_impl {
+    // Generate with no sync type conversion
+    (
+        impl Action$(<$lt:lifetime>)? for $action:ty {
+            type Future = $f_ty:ident;
+            async fn execute($($args:ident)+) -> $out:ty $code:block
+        }
+    ) => {
+        crate::action::action_impl_inner! {
+            $action => $f_ty;
+            $($lt)?;
+            async fn($($args)+) -> $out $code
+        }
+
+        #[cfg(any(feature = "sync", feature = "tokio-sync"))]
+        impl$(<$lt>)? $action {
+            /// Synchronously execute this action.
+            pub fn run(self) -> $out {
+                crate::runtime::block_on(std::future::IntoFuture::into_future(self))
+            }
+        }
+    };
+    // Generate with a sync type conversion
```

Comment:
Having to duplicate the definition here is annoying, but (AFAICT) necessary - while you can define optional matchers for a macro, you can't explicitly branch on empty or provide default values.  This means that to handle both the no-conversion and conversion case two distinct match arms are needed.  I reduced the duplication by moving the non-varying part to `action_impl_inner`.